Jazzy license: Overlay and Popover components

modernserf opened the pull request:
Overlay: http://jazzy-license.glitch.me/#StoryOverlay
Popover: http://jazzy-license.glitch.me/#StoryPopover
Shared title and sections: http://jazzy-license.glitch.me/#StoryBlocks_for_Overlay_and_Popover_content

@clottman am I handling focus & keyboard navigation for these correctly?
@sheridanvk is this sufficient to work with?

clottman commented:
on focus - when the popover/overlay closes, focus should return to the element that opened the popover/overlay. 
  - WCAG success criteria 2.4.3 (Level A): fails if focus does not return to the element after revealed content is dismissed
https://www.w3.org/WAI/WCAG21/Techniques/failures/F85

Focus/keyboard nav within the popover/overlay looks good.

clottman commented on lib/popover.js:
```diff
@@ -0,0 +1,311 @@
+  // see https://www.w3.org/TR/wai-aria-practices-1.1/#menubutton
+  return (
+    <PopoverContainer data-module="PopoverMenu" ref={rootRef} onToggle={onToggle} open={open} role="menu" {...props}>
+      <PopoverToggle aria-haspopup="menu" aria-expanded={open}>
```
Can't find the implementation of this but - why is it a `<details>` element? It should be a button.

sheridanvk replied:
hmm is that strictly true? a details element is a fully accessible way of revealing more content on the page, which seems to be the definition of a popover

modernserf replied:
This is a technique I learned from [details on details](https://docs.google.com/presentation/d/1hvnPpsJo44BTPfJx28CV95vqk_dt6na1awUbk0kmZYM/edit#slide=id.g3e31444916_1_1) [[notes](https://github.com/muan/details-on-details)]

clottman replied:
Around 25% of screenreader users use Internet explorer as their primary browser, and `<details>` isn't supported at all in IE. Some sources suggest that means the details element contents will be constantly visible as expanded in IE.

I tried testing what it looks like in IE, but the app is using generators which aren't supported either so the whole page doesn't load at all.

Are we dropping support for IE?

modernserf replied:
As far as I know, neither the editor nor community work properly in IE.

That said, if a consumer of this library wants to support IE, it can transpile this library via Babel and use this [polyfill](https://github.com/javan/details-element-polyfill) for the `<details>` element. 

Browser support strategies are [discussed a bit](https://github.com/FogCreek/shared-components/blob/master/README.md#browser-support) in the readme.

clottman replied:
What are we gaining by requiring a polyfill for the details element vs just using a button for this? 

The WAI-ARIA practices recommend using an anchor or a button element for a menu button and don't mention using `<details>` at all. I trust their recommendations since they've generally done more cross-browser and cross-AT testing.

clottman commented on lib/popover.js:
```diff
@@ -0,0 +1,311 @@
+
+  // see https://www.w3.org/TR/wai-aria-practices-1.1/#menubutton
+  return (
+    <PopoverContainer data-module="PopoverMenu" ref={rootRef} onToggle={onToggle} open={open} role="menu" {...props}>
```
Using `role="menu'` here implies also needing to set some roles on the children that are missing. 
The menu should also have aria-label. 
https://www.w3.org/TR/wai-aria-practices-1.1/#menu has details

It's not clear to me though that our popover containers are generally menus - the sign-in pop is for example but the delete project popover isn't really.

modernserf replied:
should a popover always have an aria-label, or _only_ if it has role="menu"?

clottman replied:
Only if it has role='menu'

modernserf commented:
- updated Popover to focus toggle button when its closed
- added `useOverlay` hook to manage overlay state & focus (since we can't assume toggle button and overlay are a single component)
- removed role="menu" and ARIA props that aren't intrinsic to all popovers; those can be added as needed

modernserf commented:
- removed summary/details elements
- replaced `label` prop with `renderLabel` prop

clottman commented on lib/popover.js:
```diff
@@ -87,7 +87,7 @@ const PopoverContent = ({ align, children, ...props }) => {
 };
 
-const PopoverContainer = styled.details`
+const PopoverContainer = styled.span`
```
This is wrapping all the popover contents (like buttons, headers, paragraphs etc) right? If so it should be a div or maybe `section` - span can only contain phrasing content, not flow content https://developer.mozilla.org/en-US/docs/Web/HTML/Element/span

ehmorris commented:
This may be more appropriate for followup, but I don't think we should be inserting any popovers or overlays adjacent to their activating buttons, nested in the DOM. These elements are meant to overlay basically everything, and ensuring they're overlaying properly is impossible when they could be rendering in any z-index context.

I recently fixed an overlay z-index bug in this PR https://github.com/FogCreek/Glitch-Editor/pull/106 by rendering the overlay at the document root via Portals. Overlays are easy because they're always positioned the same way, but popovers are tough because you have to position them relative to their activating button and the viewport.

Maybe a low-impact way of addressing this via shared-components is to create a component similar to this: https://github.com/FogCreek/Glitch-Editor/pull/106/files#diff-5461c4ffbf393f354ac49d2889760694 but call it "render at root" or something, and somehow remove the DOM selector dependency.

modernserf commented:
@ehmorris Overlay components do not need to be adjacent their activating buttons. 

For Popover components, you could do something like:

```js
<Popover ref={popoverRef} renderLabel={props => <Button {...props}>Show Menu</Button>}>
  {(props) => (
    <RenderAtRoot>
      <Menu {...props} positionUnderButton={popoverRef} />
    </RenderAtRoot>
  )}
</Popover>
```
I am concerned that doing this by default for every popover would have serious performance issues, since every popover would have an event listener that's updating whenever the page changes, but I can imagine circumstances where this would be necessary.

keithk commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in [`v0.19.0`](https://github.com/glitchdotcom/shared-components/releases/tag/v0.19.0) :rocket:
